apps/openchallenges/edam-etl/src/main.py

```diff
@@ -1,16 +1,100 @@
-"""Extract, transform and load EDAM concepts"""
-
+import pandas as pd
+import requests
 from os import getenv
+from typing import Optional
+import regex as re
 
 # Get config from the environment variables
-oc_db_url = getenv("OC_DB_URL")
-print(f"OC DB URL: {oc_db_url}")
 
-# TODO Download EDAM concepts from GitHub or S3 bucket (CSV file)
-print("[TODO] Download EDAM concepts from GitHub or S3 bucket (CSV file)")
+OC_DB_URL = getenv("OC_DB_URL")
+VERSION = getenv("VERSION")
+print(f"EDAM Version: {VERSION}")
+print(f"OC DB URL: {OC_DB_URL}")
+
+
+def download_edam_csv(url: str, version: str) -> Optional[bool]:
+    """Download EDAM concepts from GitHub or S3 bucket (CSV file)"""
+    print("Downloading the EDAM concepts from GitHub (CSV file)...")
+    try:
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an exception for bad response
+        with open(f"EDAM_{version}.csv", "wb") as f:
+            f.write(response.content)
+        print("EDAM concepts downloaded successfully.")
+        return True
+    except requests.RequestException as e:
+        print(f"Error downloading EDAM concepts: {e}")
+        return None
+
+
+def transform_to_dataframe(version: str) -> pd.DataFrame | None:
+    """Transform the CSV to a DataFrame with indices starting from 1"""
+    print("Processing the EDAM concepts...")
+    try:
+        df = (
+            pd.read_csv(f"EDAM_{version}.csv", usecols=["Class ID", "Preferred Label"])
+            .rename(
+                columns={"Class ID": "class_id", "Preferred Label": "preferred_label"}
+            )
+            .reset_index(drop=True)
+            .reset_index(drop=False)
+            .rename(columns={"index": "id"})
+        )
+        print("EDAM concepts processed successfully.")
+        return df
+    except FileNotFoundError as e:
+        print(f"Error processing EDAM concepts: {e}")
+        return None
+
+
+def print_info_statistics(df: pd.DataFrame) -> None:
+    """Gather data about the EDAM ontology"""
+    if df is not None:
+        print(f"Number of Concepts Transformed: {len(df)}")
+        print(f"Column names: {df.columns.tolist()}")
+
+        # Define regex patterns for each concept
+        data_pattern = re.compile(r"data_", re.IGNORECASE)
+        operation_pattern = re.compile(r"operation_", re.IGNORECASE)
+        format_pattern = re.compile(r"format_", re.IGNORECASE)
+
+        # Initialize counts
+        data_count = 0
+        operation_count = 0
+        format_count = 0
+        other_count = 0
+
+        # Loop through the class_id column and count occurrences
+        for value in df["class_id"]:
```

Comment:
A loop works!  It's very clear and straightforward on what the code is trying to do.

That being said, I think it could still be handy to take advantage of the functional programming that pandas offer, especially since `df` is a data frame.  Here's how I would use `contains()` in lieu of looping through the rows:

```python
>>> # Approach 1
>>> df["Class ID"].str.contains(r"data_\d+$", regex=True).value_counts()[True]
1493
>>>
>>> # Approach 2
>>> len(df[df["Class ID"].str.contains(r"data_\d+$", regex=True)])
1493
```

Reply:
Here are the performance differences of the three approaches:

```python
>>> # For loop
>>> timeit(
... 'for value in df["Class ID"]: data_count += 1 if data_pattern.search(value) else 0', 
... setup='import pandas as pd; import re; df = pd.read_csv("~/Downloads/EDAM.csv", usecols=["Class ID", "Preferred Label"]).fillna(""); data_pattern = re.compile(r"data_", re.IGNORECASE); data_count = 0',
... number=1000)
1.7666597969946451
>>>
>>> # using contains with value_counts
>>> timeit(
... 'df["Class ID"].str.contains(r"data_\d+$", regex=True).value_counts()[True]',
... setup='import pandas as pd; df = pd.read_csv("~/Downloads/EDAM.csv", usecols=["Class ID", "Preferred Label"]).fillna("")',
... number=1000)
1.2751378359971568
>>>
>>> # using contains with len
>>> timeit(
... 'len(df[df["Class ID"].str.contains(r"data_\d+$", regex=True)])',
... setup='import pandas as pd; df = pd.read_csv("~/Downloads/EDAM.csv", usecols=["Class ID", "Preferred Label"]).fillna("")',
... number=1000)
1.2614918049948756
```

So, not a _huge_ improvement on time, but worth noting.

Reply:
Ultimately, I say go with what makes the most sense to you!

Reply:
Honestly, I was just trying my best to incorporate both of your feedback w/o causing any additional collisions based on how I implemented the logic.